proof/compress-p2.saw

```diff
@@ -0,0 +1,216 @@
+    // need it anyway so this spec only specifies the sign_F_p_2 bit.
+    ret <- crucible_fresh_var "ret" limb_type;
+    crucible_postcond {{ (ret!1) == (sign_F_p_2 affinified.1) }};
+    crucible_return (crucible_term ret);
+  };
+};
+
+// Theorems for reasoning about `one`
+one_mont_px_abs_thm <-
+  prove_cryptol (rewrite (cryptol_ss ())
+       {{fp2_abs [[ 0x760900000002fffd, 0xebf4000bc40c0002, 0x5f48985753c758ba
+                  , 0x77ce585370525745, 0x5c071a97a256ec6d, 0x15f65ec3fa80e493],
+                  [0, 0, 0, 0, 0, 0]]
+          == Fp_2.field_unit }}) [];
+one_mont_px_invariant_thm <-
+  prove_cryptol (rewrite (cryptol_ss ())
+    {{fp2_invariant [[ 0x760900000002fffd, 0xebf4000bc40c0002, 0x5f48985753c758ba
+                     , 0x77ce585370525745, 0x5c071a97a256ec6d, 0x15f65ec3fa80e493],
+                     [0, 0, 0, 0, 0, 0]]
+       == True}}) [];
+
+// Prove POINTonE2_Compress_BE matches the spec for projective points
+POINTonE2_Compress_BE_projective_ov <- custom_verify "POINTonE2_Compress_BE"
+  [ POINTonE2_from_Jacobian_ov
+  , POINTonE2_affine_Compress_BE_ov
+  , vec_is_equal_fp2_ov ]
+  (POINTonE2_Compress_BE_spec false)
+  do {
+     // Unfolding needed for fp rewrites to apply
+     unfolding [ "POINTonE2_affine_abs"
+               , "POINTonE2_affine_rep"
+               , "POINTonE2_invariant" ];
+     (apply_fp_rewrites);
+     simplify (addsimps [ one_mont_px_abs_thm
+                        , one_mont_px_invariant_thm
+                        , seq_ite_2_thm
+                        , unrip_thm_2 ]
+                        fp2_simpset);
+     w4_unint_z3 [ "POINTonE2_abs"
+                 , "POINTonE2_invariant"
+                 , "Fp_2"
+                 , "fp2_abs"
+                 , "I2OSP48"
+                 , "fp2_invariant"
+                 , "POINTonE2_affine_invariant"];
+  };
+
+// Prove POINTonE2_Compress_BE matches the spec for affine points
+POINTonE2_Compress_BE_affine_ov <- custom_verify "POINTonE2_Compress_BE"
+  [POINTonE2_affine_Compress_BE_ov, vec_is_equal_fp2_ov]
+  (POINTonE2_Compress_BE_spec true)
+  do {
+     // Unfolding needed for fp rewrites to apply
+     unfolding [ "POINTonE2_affine_abs"
+               , "POINTonE2_affine_rep"
+               , "POINTonE2_invariant" ];
+     simplify (addsimp one_mont_px_abs_thm fp2_simpset);
+     w4_unint_z3 [ "POINTonE2_abs"
+                 , "POINTonE2_invariant"
+                 , "Fp_2"
+                 , "fp2_abs"
+                 , "I2OSP48"];
+  };
+
+let blst_p2_compress_spec is_point_affine = do {
+  out_ptr <- crucible_alloc (llvm_array 96 (llvm_int 8));
+  (inp, inp_ptr) <- ptr_to_fresh_readonly "in" POINTonE2_type;
+  crucible_precond {{ POINTonE2_invariant inp }};
+
+  if is_point_affine then do {
+    // Point is affine, so Z must be 1
+    crucible_precond {{ (fp2_abs (inp.2)) == Fp_2.field_unit }};
+    crucible_precond {{ ~(Fp_2.is_equal (fp2_abs (inp.2), Fp_2.field_zero)) }};
```

Comment:
Why is this necessary? If Z=1 then it can't be 0.

Reply:
I removed this and used a theorem when needed instead.  I also made the same change in the compress_p1 proof.